File: ts/project/.sdk/tm/elixir/lib/projectname/feature/netsim.ex
# ProjectName SDK netsim feature
#
# Network behaviour simulation. Wraps the active transport (the live
# fetch or the `test` feature's in-memory mock) and injects realistic
# network conditions so offline unit tests can exercise slowness,
# transient failures, rate limiting and outages deterministically.
#
# Every injection mode is counter-driven (per client instance) so tests
# are reproducible without mocking timers. `failRate` adds optional
# pseudo-random failures via a seeded LCG for coverage-style testing.

defmodule ProjectName.Feature.Netsim do
  alias Voxgig.Struct, as: S
  alias ProjectName.Helpers, as: H
  alias ProjectName.{Feature, Context}

  def new do
    f = Feature.base("netsim")
    S.setprop(f, "calls", 0)
    S.setprop(f, "seed", 1)
    Feature.install(f, "init", fn ctx, opts -> init(f, ctx, opts) end)
    f
  end

  def init(f, ctx, options) do
    active = Feature.init_common(f, ctx, options)

    seed_opt = S.getprop(Feature.opts(f), "seed")

    seed =
      if is_number(seed_opt) and not is_boolean(seed_opt) and trunc(seed_opt) != 0 do
        trunc(seed_opt)
      else
        1
      end

    S.setprop(f, "seed", seed)

    if active do
      utility = S.getprop(ctx, "utility")
      inner = S.getprop(utility, "fetcher")

      S.setprop(utility, "fetcher", fn fctx, fullurl, fetchdef ->
        simulate(f, fctx, fullurl, fetchdef, inner)
      end)
    end

    nil
  end

  defp simulate(f, ctx, url, fetchdef, inner) do
    opts = Feature.opts(f)
    call = S.getprop(f, "calls") + 1
    S.setprop(f, "calls", call)

    # Record the simulated conditions for test/debug inspection.
    applied = S.jm([])

    cond do
      # Total outage: every call fails at the transport level.
      S.getprop(opts, "offline") == true ->
        Feature.sleep(f, pick_latency(f))
        S.setprop(applied, "offline", true)
        track(f, ctx, applied)

        {nil,
         Context.make_error(
           ctx,
           "netsim_offline",
           "Simulated network offline (URL was: \"" <> url <> "\")"
         )}

      # Connection-level errors for the first N calls (e.g. ECONNRESET).
      call <= int_or0(S.getprop(opts, "errorTimes")) ->
        Feature.sleep(f, pick_latency(f))
        S.setprop(applied, "error", true)
        track(f, ctx, applied)

        {nil,
         Context.make_error(
           ctx,
           "netsim_conn",
           "Simulated connection error (call " <> Integer.to_string(call) <> ")"
         )}

      # Rate-limit responses (HTTP 429 + Retry-After) for the first N calls.
      call <= int_or0(S.getprop(opts, "rateLimitTimes")) ->
        Feature.sleep(f, pick_latency(f))
        S.setprop(applied, "rateLimited", true)
        track(f, ctx, applied)

        ra = S.getprop(opts, "retryAfter")
        ra = if ra == nil, do: 0, else: ra

        ra_str =
          cond do
            is_binary(ra) -> ra
            is_number(ra) -> to_string(ra)
            true -> S.stringify(ra)
          end

        respond(ctx, 429, nil,
          S.jm(["statusText", "Too Many Requests", "headers", S.jm(["retry-after", ra_str])]))

      # Retryable failure status for the first N calls, or every Nth call.
      true ->
        fs = S.getprop(opts, "failStatus")
        fail_status = if fs == nil, do: 503, else: fs
        fail_every = int_or0(S.getprop(opts, "failEvery"))
        fr = S.getprop(opts, "failRate")
        fail_rate = if pytrue?(fr), do: fr, else: 0
        fail_by_count = call <= int_or0(S.getprop(opts, "failTimes"))
        fail_by_every = fail_every > 0 and rem(call, fail_every) == 0
        fail_by_rate = fail_rate > 0 and rand(f) < fail_rate

        if fail_by_count or fail_by_every or fail_by_rate do
          Feature.sleep(f, pick_latency(f))
          S.setprop(applied, "failStatus", fail_status)
          track(f, ctx, applied)
          respond(ctx, fail_status, nil, S.jm(["statusText", "Simulated Failure"]))
        else
          # Otherwise: apply latency then delegate to the real transport.
          latency = pick_latency(f)
          S.setprop(applied, "latency", latency)
          track(f, ctx, applied)
          Feature.sleep(f, latency)
          inner.(ctx, url, fetchdef)
        end
    end
  end

  # Latency in ms: a fixed number, or a uniform sample from {min,max}.
  defp pick_latency(f) do
    latency = S.getprop(Feature.opts(f), "latency")

    cond do
      latency == nil ->
        0

      is_number(latency) and not is_boolean(latency) ->
        if latency < 0, do: 0, else: latency

      not S.ismap(latency) ->
        0

      true ->
        mn = int_or0(S.getprop(latency, "min"))
        maxv = S.getprop(latency, "max")
        mx = if maxv == nil, do: mn, else: int_or0(maxv)
        if mx <= mn, do: mn, else: mn + trunc(rand(f) * (mx - mn))
    end
  end

  # Deterministic 0..1 pseudo-random via a linear congruential generator.
  defp rand(f) do
    seed = Bitwise.band(S.getprop(f, "seed") * 1103515245 + 12345, 0x7FFFFFFF)
    S.setprop(f, "seed", seed)
    seed / 0x7FFFFFFF
  end

  defp track(f, ctx, applied) do
    client = S.getprop(f, "client")
    t = Feature.track_node(client, "_netsim", S.jm(["calls", 0, "applied", S.jt([])]))
    S.setprop(t, "calls", S.getprop(t, "calls") + 1)
    Feature.list_push(S.getprop(t, "applied"), applied)

    ctrl = S.getprop(ctx, "ctrl")
    explain = if ctrl != nil, do: S.getprop(ctrl, "explain"), else: nil
    if S.ismap(explain), do: S.setprop(explain, "netsim", t)
    nil
  end

  # Build a transport-shaped response (matching the test feature's mock)
  # with a lower-cased header map the result pipeline understands.
  defp respond(_ctx, status, data, extra) do
    out =
      S.jm([
        "status", status,
        "statusText", "OK",
        "json", fn -> data end,
        "body", "not-used"
      ])

    if S.ismap(extra) do
      Enum.each(H.entries(extra), fn {k, v} -> S.setprop(out, k, v) end)
    end

    headers = S.getprop(out, "headers")
    headers = if S.ismap(headers), do: headers, else: S.jm([])

    lower = S.jm([])
    Enum.each(H.entries(headers), fn {k, v} -> S.setprop(lower, String.downcase(to_string(k)), v) end)
    S.setprop(out, "headers", lower)

    {out, nil}
  end

  # Mirrors Python `int(x or 0)`: falsey -> 0, then integer-truncate.
  defp int_or0(v) do
    cond do
      is_integer(v) -> v
      is_number(v) -> trunc(v)
      is_binary(v) -> (case Integer.parse(v) do
                         {n, _} -> n
                         _ -> 0
                       end)
      true -> 0
    end
  end

  # Python truthiness (nil / false / "" / 0 are falsey).
  defp pytrue?(v), do: not (v == nil or v == false or v == "" or v == 0)
end

File: ts/project/.sdk/tm/zig/feature/netsim.zig
// Network behaviour simulation (mirrors go feature/netsim_feature.go / rust
// feature/netsim.rs). Wraps the active transport and injects realistic
// network conditions so offline unit tests can exercise slowness, transient
// failures, rate limiting and outages deterministically. Every mode is
// counter-driven per client; `failRate` uses a seeded LCG (gotcha #6):
//   seed = (seed*1103515245 + 12345) & 0x7fffffff.
// Error codes are exactly `netsim_offline` / `netsim_conn`.

const std = @import("std");
const h = @import("../core/helpers.zig");
const err = @import("../core/error.zig");
const types = @import("../core/types.zig");
const sup = @import("support.zig");

const Value = h.Value;
const Context = types.Context;
const Feature = types.Feature;
const Fetcher = types.Fetcher;

pub const NetsimTrack = struct {
    seed: i64 = 0,
    calls: i64 = 0,
    applied: std.ArrayList(Value),
};

pub const NetsimFeature = struct {
    name: []const u8 = "netsim",
    active: bool = true,
    add_opts: Value = .{ .null = {} },
    options: Value = .{ .null = {} },
    track: *NetsimTrack,

    pub fn make() Feature {
        const self = h.A().create(NetsimFeature) catch unreachable;
        const track = h.A().create(NetsimTrack) catch unreachable;
        track.* = .{ .applied = std.ArrayList(Value).init(h.A()) };
        self.* = .{ .track = track };
        return .{ .ptr = @ptrCast(self), .vtable = &vtable };
    }

    fn self_of(p: *anyopaque) *NetsimFeature {
        return @ptrCast(@alignCast(p));
    }

    fn vname(p: *anyopaque) []const u8 {
        return self_of(p).name;
    }
    fn vactive(p: *anyopaque) bool {
        return self_of(p).active;
    }
    fn vaddopts(p: *anyopaque) Value {
        return self_of(p).add_opts;
    }
    fn vinit(p: *anyopaque, ctx: *Context, options: Value) void {
        const self = self_of(p);
        self.options = options;
        self.active = sup.fopt_bool(options, "active", false);

        self.track.seed = sup.fopt_int(options, "seed", 0);
        if (self.track.seed == 0) self.track.seed = 1;

        if (!self.active) return;

        const util = ctx.util();
        const w = h.A().create(WrapCtx) catch unreachable;
        w.* = .{ .inner = util.fetcher, .options = options, .track = self.track };
        util.fetcher = .{ .ctx = @ptrCast(w), .call = wrapCall };
    }
    fn vdispatch(p: *anyopaque, name: []const u8, ctx: *Context) void {
        _ = p;
        _ = name;
        _ = ctx;
    }

    const vtable = Feature.VTable{
        .name = vname,
        .active = vactive,
        .add_options = vaddopts,
        .init = vinit,
        .dispatch = vdispatch,
    };
};

const WrapCtx = struct {
    inner: Fetcher,
    options: Value,
    track: *NetsimTrack,
};

fn wrapCall(p: *anyopaque, ctx: *Context, url: []const u8, fetchdef: Value) err.E!Value {
    const w: *WrapCtx = @ptrCast(@alignCast(p));
    return simulate(w.track, w.options, ctx, url, fetchdef, w.inner);
}

fn lcg_rand(track: *NetsimTrack) f64 {
    track.seed = (track.seed *% 1103515245 +% 12345) & 0x7fffffff;
    return @as(f64, @floatFromInt(track.seed)) / @as(f64, @floatFromInt(@as(i64, 0x7fffffff)));
}

fn pick_latency(track: *NetsimTrack, options: Value) i64 {
    const l = h.getp(options, "latency");
    if (h.is_noval(l)) return 0;
    if (l == .object) {
        const min = sup.fopt_int(l, "min", 0);
        const max = sup.fopt_int(l, "max", min);
        if (max <= min) return min;
        return min + @as(i64, @intFromFloat(lcg_rand(track) * @as(f64, @floatFromInt(max - min))));
    }
    return @max(sup.fopt_int(options, "latency", 0), 0);
}

fn track_applied(track: *NetsimTrack, ctx: *Context, applied: Value) void {
    track.applied.append(applied) catch {};
    const calls = track.calls;
    const applied_list = h.olist();
    for (track.applied.items) |a| applied_list.array.append(a) catch {};

    const c = ctx.ctrl;
    if (c.has_explain()) {
        h.setp(c.explain, "netsim", h.jo(&.{
            .{ "calls", h.vnum(calls) },
            .{ "applied", applied_list },
        }));
    }
}

fn respond(status: i64, data: Value, extra: []const h.Pair) Value {
    const out = h.jo(&.{
        .{ "status", h.vnum(status) },
        .{ "statusText", h.vstr("OK") },
        .{ "json", h.json_thunk(data) },
        .{ "body", h.vstr("not-used") },
        .{ "headers", h.omap() },
    });
    for (extra) |kv| h.setp(out, kv[0], kv[1]);
    return out;
}

fn simulate(track: *NetsimTrack, opts: Value, ctx: *Context, url: []const u8, fetchdef: Value, inner: Fetcher) err.E!Value {
    track.calls += 1;
    const call = track.calls;

    // Total outage: every call fails at the transport level.
    if (sup.fopt_bool(opts, "offline", false)) {
        sup.fopt_sleep(opts, pick_latency(track, opts));
        track_applied(track, ctx, h.jo(&.{.{ "offline", h.vbool(true) }}));
        return ctx.fail("netsim_offline", std.fmt.allocPrint(h.A(), "Simulated network offline (URL was: \"{s}\")", .{url}) catch "offline");
    }

    // Connection-level errors for the first N calls.
    if (call <= sup.fopt_int(opts, "errorTimes", 0)) {
        sup.fopt_sleep(opts, pick_latency(track, opts));
        track_applied(track, ctx, h.jo(&.{.{ "error", h.vbool(true) }}));
        return ctx.fail("netsim_conn", std.fmt.allocPrint(h.A(), "Simulated connection error (call {d})", .{call}) catch "conn");
    }

    // Rate-limit responses (429 + Retry-After) for the first N calls.
    if (call <= sup.fopt_int(opts, "rateLimitTimes", 0)) {
        sup.fopt_sleep(opts, pick_latency(track, opts));
        track_applied(track, ctx, h.jo(&.{.{ "rateLimited", h.vbool(true) }}));
        const headers = h.jo(&.{.{ "retry-after", h.vstr(std.fmt.allocPrint(h.A(), "{d}", .{sup.fopt_int(opts, "retryAfter", 0)}) catch "0") }});
        return respond(429, h.vnull(), &.{
            .{ "statusText", h.vstr("Too Many Requests") },
            .{ "headers", headers },
        });
    }

    // Retryable failure status: first N calls, every Nth, or at failRate.
    const fail_status = sup.fopt_int(opts, "failStatus", 503);
    const fail_every = sup.fopt_int(opts, "failEvery", 0);
    const fail_by_count = call <= sup.fopt_int(opts, "failTimes", 0);
    const fail_by_every = fail_every > 0 and @mod(call, fail_every) == 0;
    const fail_rate = sup.fopt_num(opts, "failRate", 0.0);
    const fail_by_rate = fail_rate > 0.0 and lcg_rand(track) < fail_rate;
    if (fail_by_count or fail_by_every or fail_by_rate) {
        sup.fopt_sleep(opts, pick_latency(track, opts));
        track_applied(track, ctx, h.jo(&.{.{ "failStatus", h.vnum(fail_status) }}));
        return respond(fail_status, h.vnull(), &.{.{ "statusText", h.vstr("Simulated Failure") }});
    }

    // Otherwise: apply latency then delegate to the real transport.
    const latency = pick_latency(track, opts);
    track_applied(track, ctx, h.jo(&.{.{ "latency", h.vnum(latency) }}));
    sup.fopt_sleep(opts, latency);
    return inner.invoke(ctx, url, fetchdef);
}
